### PR TITLE
Gabor/arity

### PR DESCRIPTION
**Pull request by ggreif**
Note: a possibly wrong approach is in https://github.com/dfinity/motoko/tree/gabor/higher

### CODE REVIEW

File: test/fail/ok/issue-3318.tc.ok [hunk 1]
```diff
@@ -1,24 +1,13 @@
...
-  ((Nat, Nat)) -> ((Nat, Nat))
-issue-3318.mo:24.39-24.40: type error [M0096], expression of type
-  (Nat, Nat) -> ((Nat, Nat))
-cannot produce expected type
-  (Nat, Nat) -> (Nat, Nat)
-issue-3318.mo:33.43-33.44: type error [M0096], expression of type
-  (Nat, Nat) -> ((Nat, Nat))
-cannot produce expected type
-  ((Nat, Nat)) -> ((Nat, Nat))
+issue-3318.mo:2.13-2.14: warning [M0194], unused identifier f (delete or rename to wildcard `_` or `_f`)
```
Comment (ggreif) [inline test/fail/ok/issue-3318.tc.ok:1]: This test passes now, but should it? Needs thorough checking!

File: src/mo_frontend/typing.ml [hunk 2]
```diff
@@ -637,6 +637,11 @@ and check_typ' env typ : T.typ =
     let c, typs2 = as_codomT sort.it typ2 in
     let ts1 = List.map (check_typ env') typs1 in
     let ts2 = List.map (check_typ env') typs2 in
+     if (typ.at.left.file = "Xstdin")
```
Comment (ggreif) [inline src/mo_frontend/typing.ml:640]: remove this debug code

File: src/mo_frontend/typing.ml [hunk 3]
```diff
@@ -1884,12 +1896,24 @@ and check_exp' env0 t exp : T.typ =
     t
   | _ ->
     let t' = infer_exp env0 exp in
-    if not (T.sub t' t) then
+    let u, u' =
+      let open T in
+      if is_func t && is_func t' then
+        let transfer_arity from toh = match from, List.map normalize toh with
```
Comment (ggreif) [inline src/mo_frontend/typing.ml:1902]: maybe we shouldn't normalise here, as it exposes `Tup`, where none was before. We need the some tests, but also monomorphic ones to see if there are discrepancies.

File: test/fail/ok/inference.tc.ok [hunk 4]
```diff
@@ -73,34 +73,34 @@ to argument of type
   <V>V -> V
 to produce result of type
   ()
-because no instantiation of T__102 makes
-  <V>V -> V  <:  <U>T__102 -> U
+because no instantiation of T__105 makes
```
Comment (ggreif) [inline test/fail/ok/inference.tc.ok:76]: why did these numbers shift by 3?

File: test/fail/ok/issue-3318.tc.ok (hunk 1 above)
Comment (ggreif) [reply test/fail/ok/issue-3318.tc.ok:1]: What is issue #3318?

File: src/mo_frontend/typing.ml [hunk 5]
```diff
@@ -1884,12 +1896,24 @@ and check_exp' env0 t exp : T.typ =
...
+          | [_], _ -> [seq toh]
+          | _, [Tup ts] -> ts
+          | _, _ -> toh in
+        let s, c, tbs, ts1, ts2 = as_func t in
+        let dom, cod = transfer_arity ts1, transfer_arity ts2 in
```
Comment (ggreif) [inline src/mo_frontend/typing.ml:1907]: Checking monomorphic lambdas uses `as_domT`/`as_codomT`, so maybe we should use it here too?

**Comment by luc-blaeser** [comment]
Just doing some repo cleanup for better focussing on active work:
Closing PR as it seems inactive for a longer time. Please re-open if it is still relevant and after it has been updated with latest master changes.

**Pull request closed without merging**